### PR TITLE
Support for Slurm via snakemake

### LINKED ISSUE #132: Test running under SLURM
See https://snakemake.github.io/snakemake-plugin-catalog/plugins/executor/slurm.html which suggests for the command line you just add ``--executor slurm``:

```
$ pip install snakemake-executor-plugin-slurm
$ snakemake --executor slurm ...
```

We will want to run some tests explicitly via slurm (if available), and offer to run via slurm in the public CLI API too.

### PR DESCRIPTION
**Pull request by peterjc**
This builds on #147 (and currently includes that branch) in order to potentially support any of the snakemake executors. It would close #132.

This has required a major refactoring of how we call the snakemake code to run a workflow. Previously we used ``snakemake.api`` and ``snakemake.settings`` and this was getting quite complicated I felt. In particular trying to support snakemake profiles this way seemed impractical. Now we just call two functions in ``snakemake.cli`` which parse the command line arguments (including profile support), and then run them.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [x] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [ ] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [x] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [x] Create a new branch with a short, descriptive name
- [x] Work on this branch
  - [x] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [x] new code is commented, especially in hard-to-understand areas
  - [x] corresponding changes to documentation have been made
  - [x] tests for the change have been added that demonstrate the fix or feature works
- [x] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [x] Rebase against `origin/master`
- [x] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [x] Commit branch
- [x] Submit pull request, describing the content and intent of the pull request
- [x] Request a code review
- [x] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

**Comment by peterjc** [comment]
Note this proposes to remove ``pyani_plus/snakemake/`` and instead put code in ``pyani_plus/workflows/__init__.py`` (which is the folder containing our snakemake workflow files).

File: pyani_plus/workflows/__init__.py [hunk 1]
```diff
@@ -20,3 +20,173 @@
...
+
+from pyani_plus import db_orm
+from pyani_plus.utils import progress_columns  # rich.progress config
+
+
+class ToolExecutor(str, Enum):
+    """Enum for the executor command line argument passed to snakemake."""
+
+    local = "local"
+    slurm = "slurm"
```
Comment (peterjc) [inline pyani_plus/workflows/__init__.py:41]: In order to define a command line option with predetermined choices in typer, you need to define an enum. I felt the executor enum would be best defined next to all the code for calling snakemake.

File: pyani_plus/workflows/__init__.py [hunk 2]
```diff
@@ -20,3 +20,173 @@
...
+    else:
+        duplicates = [
+            item for item in stems if stems.count(item) > 1 and item in set(stems)
+        ]
+        msg = (
+            f"Duplicated stems found for {sorted(set(duplicates))}. Please investigate."
+        )
+        raise ValueError(msg)
+
+    return input_files
```
Comment (peterjc) [inline pyani_plus/workflows/__init__.py:66]: This is the same ``check_input_stems`` function as before, but relocated.

**Comment by peterjc** [comment]
The key point in this design is how to fine tune the HPC configuration is left to the user via standard snakemake profile configuration, e.g. this works for me:

```console
$ cat snakemake-profile-archie-west/config.yaml
executor: slurm
jobs: 10
cores: 10
precommand: "module purge && module load miniconda && conda activate pyani-plus_py312 && python --version"

default-resources:
    slurm_partition: dev
    slurm_account: pritchard-grxiv
```

Sample usage:

```console
$ rm -rf x.db .snakemake
$ export SNAKEMAKE_PROFILE=$PWD/snakemake-profile-archie-west
$ pyani-plus fastani --database x.db --create-db --name 3V tests/fixtures/viral_example/ --executor slurm
Indexing FASTAs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 3/3
Database already has 0 of 3x3=9 comparisons, 9 needed
Using profile /users/wkb24128/repositories/pyani-plus/snakemake-profile-archie-west for setting default command line arguments.
Comparing pairs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:34 9/9
Logged new run-id 1 for method fastANI with 3 genomes to database x.db
```

We might want to add something like a ``--snakemake-profile`` option and/or remove the current ``--executor slurm`` bit since that too can be in a profile I think.

File: pyani_plus/workflows/__init__.py [hunk 3]
```diff
@@ -20,3 +20,173 @@
...
+
+    # Path to anim snakemake file
+    snakefile = impresources.files(workflows) / workflow_name
+
+    # Want snakemake to pull everything else from its profiles,
+    # most easily controlled via setting $SNAKEMAKE_PROFILE
+    parser, args = parse_args(
+        [
+            "--quiet",
+            "all" if show_progress_bar else "rules",
```
Comment (peterjc) [inline pyani_plus/workflows/__init__.py:146]: Quiet by default is OK, but this needs to be controllable. For example adding a ``--verbose`` vs ``--quiet`` switch?

**Comment by peterjc** [comment]
Another example working slurm setup, tested on https://cropdiversity.ac.uk/

```
$ cat snakemake-profile-crop-diversity/config.yaml 
executor: slurm
jobs: 10
cores: 10

default-resources:
    slurm_partition: short
    slurm_account: cropdiv-acc
    runtime: 15
```

This one is simpler in that everyone uses conda by default, and to keep this very simple I installed pyani-plus and its dependencies into my base environment - so no precommand etc needed at all. If pyani-plus set a default for jobs/cores, this example might work without any configuration at all?

```
$ rm -rf x.db
$ export SNAKEMAKE_PROFILE=$PWD/snakemake-profile-crop-diversity
$ time pyani-plus fastani --database x.db --create-db --name 3V tests/fixtures/viral_example/ --executor slurm
Indexing FASTAs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:00:00 3/3
Database already has 0 of 3x3=9 comparisons, 9 needed
Using profile /home/xxx/repositories/pyani-plus/snakemake-profile-crop-diversity for setting default command line arguments.
Comparing pairs ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 100% 0:01:28 9/9
Logged new run-id 1 for method fastANI with 3 genomes to database x.db

real	3m34.333s
user	0m3.448s
sys	0m3.416s
```

Note there is some considerable overhead aside from the actual computation - hopefully this is fairly static as the number of comparisons rises. Some of this could be snakemake default waits to allow the filesystem to catch up?

**Comment by widdowquinn** [comment]
I like the direction this is going. The refactoring makes sense, and using the CLI directly through code rather than descending into the (undocumented) API is a big improvement.

**Comment by peterjc** [comment]
Superseded by #158.

**Pull request closed without merging**